neqo-transport/src/connection/tests/ecn.rs

```diff
@@ -0,0 +1,512 @@
+    for _ in 0..burst {
+        let client_pkt = send_something_with_modifier(&mut client, now, &mut orig_path_modifier);
+        server.process_input(&client_pkt, now);
+    }
+
+    if let Some(ack) = server.process_output(now).dgram() {
+        client.process_input(&ack, now);
+    }
+
+    let client_pkt = send_something(&mut client, now);
+    let tos_before_migration = client_pkt.tos();
+    server.process_input(&orig_path_modifier(client_pkt).unwrap(), now);
+
+    client
+        .migrate(Some(DEFAULT_ADDR_V4), Some(DEFAULT_ADDR_V4), false, now)
+        .unwrap();
+
+    let mut migrated = false;
+    let probe = new_path_modifier(client.process_output(now).dgram().unwrap());
+    if probe.is_some() {
+        let probe = probe.unwrap();
+        assert_v4_path(&probe, true); // Contains PATH_CHALLENGE.
+        assert_eq!(client.stats().frame_tx.path_challenge, 1);
+        let probe_cid = ConnectionId::from(get_cid(&probe));
+
+        let resp = new_path_modifier(server.process(Some(&probe), now).dgram().unwrap()).unwrap();
+        assert_v4_path(&resp, true);
+        assert_eq!(server.stats().frame_tx.path_response, 1);
+        assert_eq!(server.stats().frame_tx.path_challenge, 1);
+
+        // Data continues to be exchanged on the old path.
+        let client_data = send_something_with_modifier(&mut client, now, &mut orig_path_modifier);
+        assert_ne!(get_cid(&client_data), probe_cid);
+        assert_v6_path(&client_data, false);
+        server.process_input(&client_data, now);
+        let server_data = send_something_with_modifier(&mut server, now, &mut orig_path_modifier);
+        assert_v6_path(&server_data, false);
+        client.process_input(&server_data, now);
+
+        // Once the client receives the probe response, it migrates to the new path.
+        client.process_input(&resp, now);
+        assert_eq!(client.stats().frame_rx.path_challenge, 1);
+        migrated = true;
+
+        let migrate_client = send_something_with_modifier(&mut client, now, &mut new_path_modifier);
+        assert_v4_path(&migrate_client, true); // Responds to server probe.
+
+        // The server now sees the migration and will switch over.
+        // However, it will probe the old path again, even though it has just
+        // received a response to its last probe, because it needs to verify
+        // that the migration is genuine.
+        server.process_input(&migrate_client, now);
+    }
+
+    let stream_before = server.stats().frame_tx.stream;
+    let probe_old_server = send_something_with_modifier(&mut server, now, &mut orig_path_modifier);
+    // This is just the double-check probe; no STREAM frames.
+    assert_v6_path(&probe_old_server, migrated);
+    assert_eq!(
+        server.stats().frame_tx.path_challenge,
+        if migrated { 2 } else { 0 }
+    );
+    assert_eq!(
+        server.stats().frame_tx.stream,
+        if migrated { stream_before } else { 1 }
+    );
+
+    if migrated {
```

Comment:
In looking at the tests below, you aren't really checking that the connection migrated successfully.  I know that it's more outputs on the function, but I think it would be helpful to track if migration happened.